File: csharp/solarSystem/solarsystem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Reactive.Concurrency;
using System.Reactive;
using System.Reactive.Linq;
using System.Reactive.Disposables;
using RTI.RxDDS;

namespace solar
{
    class SolarSystem
    {
        struct PlanetInfo
        {
            public int orbitRadius;
            public int size;
            public int yearInEarthDays;

            public PlanetInfo(int orbitRadius, int size, int year)
            {
                this.orbitRadius = orbitRadius;
                this.size = size;
                this.yearInEarthDays = year;
            }
        };

        readonly static int earthYear = 365;

        DDS.TypedDataWriter<ShapeType> triangleWriter;
        DDS.TypedDataWriter<ShapeType> circleWriter;
        Dictionary<string, PlanetInfo> planets;

        static void Main(string[] args)
        {
            int domainId = 0;
            IDisposable disposable = null;

            if (args.Length >= 1)
            {
                if (!Int32.TryParse(args[0], out domainId))
                {
                    Console.WriteLine("Invalid domainId. Quitting...");
                    return;
                }
            }

            if (args.Length >= 2)
            {
                if (args[1] == "bigbang")
                    disposable = new SolarSystem(domainId).bigbang();
                else if (args[1] == "bigbang2")
                    disposable = new SolarSystem(domainId).bigbang2();
                else if (args[1] == "bigbang3")
                    disposable = new SolarSystem(domainId).bigbang3();
                else
                {
                    Console.WriteLine("Specify bigbang | bigbang2 | bigbang3. Quitting...");
                    return;
                }
            }

            for (; disposable != null; )
            {
                Console.WriteLine("Press Enter to dispose...");
                ConsoleKeyInfo info = Console.ReadKey(true);
                if (info.Key == ConsoleKey.Enter)
                {
                    disposable.Dispose();
                    break;
                }
            }
        }

        SolarSystem(int domainId)
        {
            planets =
                new Dictionary<string, PlanetInfo>();

            // PlanetInfo(orbitRadius, size, year-in-earth-days)
            planets.Add("Mercury", new PlanetInfo(30,   8,  88));
            planets.Add("Venus",   new PlanetInfo(50,  15,  225));
            planets.Add("Earth",   new PlanetInfo(70,  17,  365));
            planets.Add("Mars",    new PlanetInfo(90,  12,  686));
            planets.Add("Jupiter", new PlanetInfo(120, 30, 4329));
            
            // Yes, the Moon too!
            planets.Add("Moon",    new PlanetInfo(20,  8,  365));

            RTI.RxDDS.DefaultParticipant.DomainId = domainId;

            RTI.RxDDS.DefaultParticipant
                     .RegisterType<ShapeType, 
                                   ShapeTypeTypeSupport>();

            this.triangleWriter =
                RTI.RxDDS.DefaultParticipant
                         .CreateDataWriter<ShapeType>("Triangle");

            this.circleWriter =
                RTI.RxDDS.DefaultParticipant
                         .CreateDataWriter<ShapeType>("Circle");
        }

        ~SolarSystem()
        {
            RTI.RxDDS.DefaultParticipant.Shutdown();
        }        

        ShapeType planetLocation(ShapeType sun, double degree, string planetName)
        {
            var planet = planets[planetName];
            return new ShapeType
            {
                x = sun.x + (int)(planet.orbitRadius * Math.Cos(degree * Math.PI / 180)),
                y = sun.y + (int)(planet.orbitRadius * Math.Sin(degree * Math.PI / 180)),
                color = sun.color,
                shapesize = planet.size
            };
        }

        IDisposable bigbang()
        {
            // The Sun was born here!
            var sunOrbit =
               DDSObservable.FromTopic<ShapeType>(DefaultParticipant.Instance,
                                                  "Square",
                                                  Scheduler.Default);

            // Earth was born here!
            double earthDegree = 0;
            IObservable<ShapeType> earthOrbit =
                    sunOrbit.Select(sunLoc =>
                    {
                        earthDegree+=2;
                        return planetLocation(sunLoc, earthDegree, "Earth");
                    })
                    .Publish()
                    .RefCount();

            // The Moon was born here!
            double moonDegree = 0;
            IObservable<ShapeType> moonOrbit
                = earthOrbit.Select(earthLoc =>
                {
                    moonDegree += 7;
                    return planetLocation(earthLoc, moonDegree, "Moon");
                });

            var assumedInstance = new ShapeType { color = "BLUE" };
            // begin orbiting!
            return new CompositeDisposable(new IDisposable[] { 
                      earthOrbit.DisposeAtEnd(circleWriter, assumedInstance).Subscribe(),
                      moonOrbit.DisposeAtEnd(triangleWriter, assumedInstance).Subscribe()
                  });
        }

        double output(string msg, long t, double degree)
        {
            Console.WriteLine(msg, t, degree);
            return degree;
        }

        IObservable<ShapeType>
                planetOrbitLinq(IObservable<long> ticks,
                                IObservable<ShapeType> sunOrbit,
                                string planetName)
        {
            // This implementation does not propagate dispose
            IObservable<ShapeType> planetOrbit =
                    from t in ticks
                    from sunLoc in sunOrbit.Take(1)
                    let degree = (int)(t * earthYear / planets[planetName].yearInEarthDays)
                    let op = output("t = {0}, degree = {1}", t, degree)
                    select planetLocation(sunLoc, degree, planetName);

            return planetOrbit;
        }

        // Space and Time were born here!
        IDisposable bigbang2()
        {
            // The beginning of time!
            var ticks =
               Observable.Interval(TimeSpan.FromMilliseconds(20),
                                   new EventLoopScheduler());

            // The Sun was born here!
            var sunOrbit =
               DDSObservable.FromTopic<ShapeType>(DefaultParticipant.Instance,
                                                  "Square",
                                                  Scheduler.Default);

            // planets were born here!
            var mercuryOrbit = planetOrbitLinq(ticks, sunOrbit, "Mercury");
            var venusOrbit   = planetOrbitLinq(ticks, sunOrbit, "Venus");
            var earthOrbit   = planetOrbitLinq(ticks, sunOrbit, "Earth").Publish().RefCount();
            var marsOrbit    = planetOrbitLinq(ticks, sunOrbit, "Mars");
            var jupiterOrbit = planetOrbitLinq(ticks, sunOrbit, "Jupiter");

            // The Moon was born here!
            double moonDegree = 0;
            var moonOrbit
                = earthOrbit.Select(earthLoc =>
                {
                    moonDegree += 7;
                    return planetLocation(earthLoc, moonDegree, "Moon");
                });

            // begin orbiting but be ready for the big crunch!
            return new CompositeDisposable(new IDisposable[] { 
                      mercuryOrbit.Subscribe(circleWriter),
                      venusOrbit.Subscribe(circleWriter),
                      earthOrbit.Subscribe(circleWriter),
                      marsOrbit.Subscribe(circleWriter),
                      jupiterOrbit.Subscribe(circleWriter),
                      moonOrbit.Subscribe(triangleWriter)
                  });
        }

        IObservable<ShapeType>
        planetOrbit(IObservable<long> ticks,
                    IObservable<ShapeType> sunOrbit,
                    string planetName)
        {/*
            return ticks.Select(t =>
            {
                var degree = (int)(t * earthYear / planets[planetName].yearInEarthDays);
                Console.WriteLine("t = {0}, degree = {1}", t, degree);
                return sunOrbit.Select(sunLoc => planetLocation(sunLoc, degree, planetName));
            })
            .Switch()
            .Do(planet => Console.WriteLine("x = {0}, y = {1}",  planet.x, planet.y)); ;
            */return planetOrbitLinq(ticks, sunOrbit, planetName);
        }

        IDisposable bigbang3()
        {
            var ticks =
               Observable.Interval(TimeSpan.FromMilliseconds(20),
                                   new EventLoopScheduler());

            var sunOrbitStream =
               DDSObservable.FromKeyedTopic<string, ShapeType>(
                                  DefaultParticipant.Instance,
                                  "Square",
                                  shape => shape.color,
                                  Scheduler.Default);

            return sunOrbitStream.Select(sunOrbit =>
            {
                //var mercuryOrbit = planetOrbit(ticks, sunOrbit, "Mercury");
                //var venusOrbit   = planetOrbit(ticks, sunOrbit, "Venus");
                var earthOrbit   = planetOrbit(ticks, sunOrbit, "Earth").Publish().RefCount();
                //var marsOrbit    = planetOrbit(ticks, sunOrbit, "Mars");
                //var jupiterOrbit = planetOrbit(ticks, sunOrbit, "Jupiter");

                double moonDegree = 0;
                var moonOrbit
                    = earthOrbit.Select(earthLoc =>
                    {
                        moonDegree += 3;
                        Console.WriteLine("moonDegree = {0}", moonDegree);
                        return planetLocation(earthLoc, moonDegree, "Moon");
                    });

                var instance = new ShapeType { color = sunOrbit.Key };
                return new CompositeDisposable(new IDisposable[] { 
                          //mercuryOrbit.DisposeAtEnd(circleWriter,   instance).Subscribe(),
                          //venusOrbit  .DisposeAtEnd(circleWriter,   instance).Subscribe(),
                          earthOrbit  .DisposeAtEnd(circleWriter,   instance).Subscribe(),
                          //marsOrbit   .DisposeAtEnd(circleWriter,   instance).Subscribe(),
                          //jupiterOrbit.DisposeAtEnd(circleWriter,   instance).Subscribe(),
                          moonOrbit   .DisposeAtEnd(triangleWriter, instance).Subscribe()
                      });

            }).Subscribe();
        }
    }
}

File: cpp/rx4dds-test/solar_system.cpp
#include "solar_system.h"
#include <cmath>

namespace {
  const double M_PI = 3.14159265358979323846;
}

std::unordered_map<std::string, PlanetInfo> SolarSystem::planets;

SolarSystem::SolarSystem(int domain_id)
  : participant_(domain_id),
    circle_topic_(participant_, "Circle"),
    triangle_topic_(participant_, "Triangle"),
    circle_writer_(dds::pub::Publisher(participant_), circle_topic_),
    triangle_writer_(dds::pub::Publisher(participant_), triangle_topic_),
    scheduler_(rxcpp::schedulers::make_current_thread()),
    worker_(scheduler_.create_worker()),
    topic_subscription_(participant_, "Square", waitset_, worker_)
{
  // PlanetInfo(orbitRadius, size, year-in-earth-days)
  planets["Mercury"] = PlanetInfo(30, 8, 88);
  planets["Venus"]   = PlanetInfo(50, 15, 225);
  planets["Earth"]   = PlanetInfo(70, 17, 365);
  planets["Mars"]    = PlanetInfo(90, 12, 686);
  planets["Jupiter"] = PlanetInfo(120, 30, 4329);

  // Yes, the Moon too! Right, it's not a planet.
  planets["Moon"] =  PlanetInfo(20, 8, 365);
}

ShapeType SolarSystem::planet_location(const ShapeType & sun, 
                                       double degree, 
                                       const std::string & planet_name) 
{
  const PlanetInfo & planet = planets.at(planet_name);
  return ShapeType(
    sun.color(),
    sun.x() + (int) (planet.orbit_radius_ * std::cos(degree * M_PI / 180)),
    sun.y() + (int) (planet.orbit_radius_ * std::sin(degree * M_PI / 180)),
    planet.size_);
}

rxcpp::composite_subscription SolarSystem::big_bang()
{
  ShapeType blue_instance("BLUE", -1, -1, -1);

  // The Sun observable
  auto sun_orbit =
    topic_subscription_.create_data_observable()
    >> rx4dds::complete_on_dispose()
    >> rx4dds::error_on_no_alive_writers()
    >> rx4dds::skip_invalid_samples()
    >> rx4dds::map_sample_to_data();

  // The Earth observable
  int earth_degree = 0;
  auto earth_orbit =
    sun_orbit.map([earth_degree](const ShapeType & sun_loc) 
    {
      const_cast<int &>(earth_degree) = (earth_degree + 2) % 360;
      return SolarSystem::planet_location(sun_loc, earth_degree, "Earth");
  })
  >> rx4dds::publish_over_dds(circle_writer_, blue_instance);

  // The Moon observable
  int moon_degree = 0;
  auto moon_orbit
    = earth_orbit.map([moon_degree](const ShapeType earth_loc)
      {
        const_cast<int &>(moon_degree) = (moon_degree + 7) % 360;
        return SolarSystem::planet_location(earth_loc, moon_degree, "Moon");
      })
      >> rx4dds::publish_over_dds(triangle_writer_, blue_instance);

  return moon_orbit.subscribe();
}

rxcpp::composite_subscription SolarSystem::big_bang2()
{
  auto solarsystem_stream =
    topic_subscription_.create_data_observable()
    >> rx4dds::group_by_dds_instance<dds::core::string, ShapeType>(
        [](const ShapeType & shape) { return shape.color(); });

  typedef
    rxcpp::grouped_observable < dds::core::string, rti::sub::LoanedSample<ShapeType> >
      GroupedShapeObservable;

  rxcpp::composite_subscription subscription =
    solarsystem_stream
    .flat_map([this](GroupedShapeObservable go)
    {
      ShapeType instance(go.get_key(), -1, -1, -1);

      auto sun_orbit =
        go  >> rx4dds::to_unkeyed()
            >> rx4dds::error_on_no_alive_writers()
            >> rx4dds::skip_invalid_samples()
            >> rx4dds::map_sample_to_data();

      int earth_degree = 0;
      auto earth_orbit =
        sun_orbit.map([earth_degree](const ShapeType & sun_loc) {
          const_cast<int &>(earth_degree) = (earth_degree + 3) % 360;
          return SolarSystem::planet_location(sun_loc, earth_degree, "Earth");
        })
        >> rx4dds::publish_over_dds(circle_writer_, instance);

      int moon_degree = 0;
      auto moon_orbit
        = earth_orbit.map([moon_degree](const ShapeType & earth_loc)
      {
        const_cast<int &>(moon_degree) = (moon_degree + 9) % 360;
        return SolarSystem::planet_location(earth_loc, moon_degree, "Moon");
      });

      return moon_orbit
        >> rx4dds::publish_over_dds(triangle_writer_, instance);
    },
    [](GroupedShapeObservable, const ShapeType &) {
    return 0;
  })
  .subscribe();

  return subscription;
}

void SolarSystem::orbit(rxcpp::composite_subscription subscription)
{
  rxcpp::schedulers::schedulable schedulable =
    rxcpp::schedulers::make_schedulable(
    worker_,
    [this](const rxcpp::schedulers::schedulable &) {
      const_cast<dds::core::cond::WaitSet &>(waitset_).dispatch(dds::core::Duration(4));
  });

  while (true) {
    worker_.schedule(schedulable);
  }
}
